src/viewport_renderer.rs

```diff
@@ -2,23 +2,132 @@ use std::collections::HashMap;
 use std::convert::TryFrom;
 use std::error;
 use std::fmt;
+use std::io;
 use std::mem;
 
 use nalgebra::base::Matrix4;
 
 use wgpu::winit;
 
+macro_rules! include_shader {
+    ($name:expr) => {{
+        include_bytes!(concat!(env!("OUT_DIR"), "/shaders/", $name))
+    }};
+}
+
 const DEPTH_FORMAT: wgpu::TextureFormat = wgpu::TextureFormat::D32Float;
 
+const SHADER_MATCAP_VERT: &[u8] = include_shader!("matcap.vert.spv");
+const SHADER_MATCAP_FRAG: &[u8] = include_shader!("matcap.frag.spv");
+
+const MATCAP_TEXTURE_BYTES: &[u8] = include_bytes!("../resources/matcap.png");
+
+/// The geometry vertex data as uploaded on the GPU.
+///
+/// Positions and normals are internally `[f32; 4]` with the last
+/// component filled in as 0.0 or 1.0 depending on it being a position
+/// or vector.
+#[repr(C)]
 #[derive(Debug, Clone, Copy, PartialEq)]
 pub struct Vertex {
-    pub position: [f32; 3],
+    // These are defined as [f32; 4] for 2 reasons:
+    //
+    // - point vs vector clarity
+    // - padding: vec3 has the same size as vec4 in std140 layout.
+    //   https://www.khronos.org/opengl/wiki/Interface_Block_(GLSL)
+    //
+    // We might decide to pack more useful data than just 0/1 in the
+    // last component later.
+    pub position: [f32; 4],
+    pub normal: [f32; 4],
 }
 
 // FIXME: @Optimization Determine u16/u32 dynamically per geometry to
 // save memory
+/// The geometry indices as uploaded on the GPU.
 pub type Index = u32;
 
+/// The geometry containing index and vertex data in the format as
+/// will be uploaded on the GPU.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Geometry {
+    indices: Option<Vec<Index>>,
+    vertex_data: Vec<Vertex>,
+}
+
+impl Geometry {
```

Comment:
This asks for a test or two.

Reply:
Added some more asserts and tests for them. I don't think this is the place to do in-depth validation, so I stopped there.